dashboard/pages/data_projection/callbacks.py

```diff
@@ -0,0 +1,306 @@
+
+PROJECTION_SETUP = [
+    (PCA(n_components=2), "PCA"),
+    (
+        UMAP(
+            n_components=2,
+            n_neighbors=10,
+            min_dist=0.1,
+        ),
+        "UMAP",
+    ),
+]
+
+# === STAGE 1 ===
+
+
+def on_projection_files_upload(
+    content: str | None,
+    filenames: list[str],
+    last_modified: int,
+    stored_uuid: str,
+    file_storage: FileStorage,
+) -> html.Div:
+    """
+    Callback for file upload. TBD
+
+    :param content: base64 encoded file content
+    :param stored_uuid: session uuid
+    :param file_storage: file storage
+    :return: icon indicating the status of the upload
+    """
+    if content is None:
+        return no_update
+    if stored_uuid is None:
+        stored_uuid = str(uuid.uuid4())
+
+    projection_files = []
+
+    for content, filename in zip(content, filenames):
+        filename, extension = filename.split(".")
+        if extension == "csv":
+            decoded = base64.b64decode(content.split(",")[1]).decode("utf-8")
+            projection_files.append((filename, io.StringIO(decoded)))
+
+    if projection_files:
+        assays_preprocessor = MergedAssaysPreprocessor()
+        assays_preprocessor.combine_assays_for_projections(tuple(projection_files))
+        assays_merged_df = assays_preprocessor.get_processed_compounds_df()
+
+    saved_name = f"{stored_uuid}_assays_merged.pq"
+    file_storage.save_file(saved_name, assays_merged_df.reset_index().to_parquet())
+
+    return html.Div(
+        children=[
+            make_file_list_component(filenames, [], 1),
+        ],
+    )
+
+
+# === STAGE 2 ===
+
+
+def on_projections_visualization_entry(
+    current_stage: int,
+    stored_uuid: str,
+    file_storage: FileStorage,
+) -> tuple[go.Figure, html.Div, html.Div, html.Div, html.Div, html.Div]:
+    """
+    Callback for projections visualization stage entry.
+    It loads the data from the storage, computes and visualizes the projections.
+
+    :param current_stage: index of the current stage
+    :param stored_uuid: session uuid
+    :param file_storage: file storage
+    :return: figure with projections, table with projections, dropdown with projection attributes,
+    dropdown with projection methods, projection info, checkbox for controls
+    """
+    if current_stage != 1:
+        return no_update
+
+    load_name = f"{stored_uuid}_assays_merged.pq"
+    merged_df = pd.read_parquet(pa.BufferReader(file_storage.read_file(load_name)))
+
+    # take only columns with projections i.e. having % in the name
+    projection_columns = [col for col in merged_df.columns if "%" in col]
+    controls_df = generate_controls(projection_columns)
+
+    assays_preprocessor = MergedAssaysPreprocessor()
+    assays_preprocessor.set_compounds_df(merged_df).set_controls_df(
+        controls_df
+    ).set_columns_for_projection(projection_columns)
+
+    for projector, name in PROJECTION_SETUP:
+        assays_preprocessor.apply_projection(projector, name)
```

Comment:
doesn't this apply the same projection twice?

Reply:
One's for `fit_transform`, the other - solely `transform` for projecting control values.
On the other hand, I can just modify `apply_projection` to make it in a single step - will do it that way to make it more intuitive